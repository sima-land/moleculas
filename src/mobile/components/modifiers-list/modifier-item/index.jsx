import React from 'react';
import classes from './modifier-item.scss';
import classnames from 'classnames/bind';
import { MODIFIERS_TYPES } from '@dev-dep/ui-nucleons/constants';
import Price from '@dev-dep/ui-nucleons/price';
import Text from '@dev-dep/ui-nucleons/text';
import CheckSVG from '@dev-dep/ui-quarks/icons/24x24/Stroked/check';

export const cx = classnames.bind(classes);

/**
 * Компонент модификатора в списке.
 * @param {Object} props Свойства компонента.
 * @param {string} props.name Наименование модификатора.
 * @param {boolean} [props.selected] Выбран ли модификатор.
 * @param {'text'|'image'|'color'} [props.type='text'] Тип модификатора.
 * @param {string} [props.color] Цвет модификатора.
 * @param {string} [props.image] URL изображения модификатора.
 * @param {number} props.price Цена товара-модификатора.
 * @param {string} props.currencyGrapheme Графема валюты пользователя.
 * @param {string} [props.additionalText] Дополнительный текст.
 * @param {Function} props.onClick Обработчик клика на модификатор.
 * @return {ReactElement} Компонент модификатора.
 */
export const ModifierItem = ({
  name,
  selected,
  type = 'text',
  color,
  image,
  price,
  currencyGrapheme,
  additionalText,
  onClick,
}) => (
  <div
    onClick={onClick}
    className={cx('wrapper', selected && 'selected')}
  >
    {Boolean(type === MODIFIERS_TYPES.image && image) && (
      <img
        className={cx('adornment')}
        src={image}
        alt={name}
      />
    )}
    {Boolean(type === MODIFIERS_TYPES.color && color) && (
      <div
        className={cx('adornment', 'color')}
        style={{ backgroundColor: color }}
      />
    )}
    <div className={cx('info')}>
      <span className={cx('name')}>{name}</span>
      {Boolean(price) && (
        <Price
          value={price}
          currencyGrapheme={currencyGrapheme}
          className={cx('price', additionalText && 'additional')}
          fractionalSign=','
        />
      )}
      {Boolean(additionalText) && (
        <Text
          color='gray38'
          children={additionalText}
          size={12}
          lineHeight={16}
        />
      )}
    </div>
    <div className={cx('icon-wrapper')}>
      {selected && (
        <div className={cx('icon')}>
          <CheckSVG
            fill='#fff'
            role='presentation'
          />
        </div>
      )}
    </div>
  </div>
);
