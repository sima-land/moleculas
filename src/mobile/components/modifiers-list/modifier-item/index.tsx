import React from 'react';
import classes from './modifier-item.module.scss';
import classnames from 'classnames/bind';
import { MODIFIERS_TYPES } from '../../../../common/constants';
import { Price } from '@sima-land/ui-nucleons/price';
import { Text } from '@sima-land/ui-nucleons/text';
import CheckSVG from '@sima-land/ui-quarks/icons/24x24/Stroked/check';

export interface ModifierItemProps {
  name: string
  selected?: boolean
  type?: 'text'|'image'|'color'
  color?: string
  image?: string
  price: number
  currencyGrapheme?: string
  additionalText?: string
  onClick: React.MouseEventHandler<HTMLDivElement>
}

export const cx = classnames.bind(classes);

/**
 * Компонент модификатора в списке.
 * @param props Свойства компонента.
 * @param props.name Наименование модификатора.
 * @param props.selected Выбран ли модификатор.
 * @param props.type Тип модификатора.
 * @param props.color Цвет модификатора.
 * @param props.image URL изображения модификатора.
 * @param props.price Цена товара-модификатора.
 * @param props.currencyGrapheme Графема валюты пользователя.
 * @param props.additionalText Дополнительный текст.
 * @param props.onClick Обработчик клика на модификатор.
 * @return Элемент.
 */
export const ModifierItem = ({
  name,
  selected,
  type = 'text',
  color,
  image,
  price,
  currencyGrapheme,
  additionalText,
  onClick,
}: ModifierItemProps) => (
  <div
    onClick={onClick}
    className={cx('wrapper', selected && 'selected')}
  >
    {Boolean(type === MODIFIERS_TYPES.image && image) && (
      <img
        className={cx('adornment')}
        src={image}
        alt={name}
      />
    )}
    {Boolean(type === MODIFIERS_TYPES.color && color) && (
      <div
        className={cx('adornment', 'color')}
        style={{ backgroundColor: color }}
      />
    )}
    <div className={cx('info')}>
      <span className={cx('name')}>{name}</span>
      {Boolean(price) && (
        <Price
          value={price}
          currencyGrapheme={currencyGrapheme}
          className={cx('price', additionalText && 'additional')}
        />
      )}
      {Boolean(additionalText) && (
        <Text
          color='gray38'
          children={additionalText}
          size={12}
          lineHeight={16}
        />
      )}
    </div>
    <div className={cx('icon-wrapper')}>
      {selected && (
        <div className={cx('icon')}>
          <CheckSVG
            fill='#fff'
            role='presentation'
          />
        </div>
      )}
    </div>
  </div>
);
