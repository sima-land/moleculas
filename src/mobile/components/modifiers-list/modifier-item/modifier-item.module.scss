@use 'node_modules/@dev-dep/ui-nucleons/colors';

.wrapper {
  box-sizing: border-box;
  cursor: pointer;
  display: flex;
  padding: 16px;
  width: 100%;
  &:not(:last-child) {
    box-shadow: inset 0 -1px 0 colors.$gray12;
  }
  &.selected {
    background: colors.$gray4;
    cursor: auto;
    pointer-events: none;
  }
}

.adornment {
  flex-shrink: 0;
  display: block;
  width: 48px;
  height: 48px;
  margin-right: 16px;
  border-radius: 4px;
  &.color {
    box-sizing: border-box;
    border: 1px solid #e0e0e0;
  }
}

.info {
  flex-grow: 1;
  display: flex;
  flex-direction: column;
  .name {
    &::first-letter {
      text-transform: capitalize;
    }
    font-size: 16px;
    line-height: 24px;
    font-weight: 500;
    padding-bottom: 4px;
  }
  .price {
    font-size: 14px;
    line-height: 20px;
    &.additional {
      padding-bottom: 12px;
    }
  }
}

.icon-wrapper {
  flex-shrink: 0;
  margin-left: 16px;
  width: 48px;
  height: 48px;
}

.icon {
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  border-radius: 50%;
  background-color: colors.$additional-teal;
}
