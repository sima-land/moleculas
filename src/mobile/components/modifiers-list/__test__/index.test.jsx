import React from 'react';
import { shallow, mount } from 'enzyme';
import { omit } from 'lodash';
import { ModifiersList } from '..';
import { ModifierItem } from '../modifier-item';
import { Link } from '@sima-land/ui-nucleons/link';
import StatementSVG from '@sima-land/ui-quarks/icons/24x24/Stroked/statement';

describe('<ModifiersList />', () => {
  const items = [
    { name: 'testModifier - 1', type: 'text', price: 123 },
    { name: 'testModifier - 2', type: 'text', price: 233 },
  ];
  it('should renders correctly without props', () => {
    const wrapper = shallow(<ModifiersList />);
    expect(wrapper.find(Link)).toHaveLength(0);
    expect(wrapper.find(StatementSVG)).toHaveLength(0);
    expect(wrapper.find(ModifierItem)).toHaveLength(0);
    expect(wrapper).toMatchSnapshot();
  });
  it('should renders correctly with correct props', () => {
    const wrapper = shallow(<ModifiersList items={items} currencyGrapheme='$' />);
    expect(wrapper.find(ModifierItem)).toHaveLength(2);
    expect(omit(wrapper.find(ModifierItem).at(0).props(), ['onClick'])).toEqual({
      ...items[0],
      currencyGrapheme: '$',
    });
    expect(omit(wrapper.find(ModifierItem).at(1).props(), ['onClick'])).toEqual({
      ...items[1],
      currencyGrapheme: '$',
    });
    expect(wrapper).toMatchSnapshot();
  });
  it('should render correct with sizesTableUrl', () => {
    const wrapper = mount(<ModifiersList sizesTableUrl='test.ru/123' />);
    expect(wrapper.find(Link)).toHaveLength(1);
    expect(wrapper.find(Link).text()).toEqual('Таблица размеров');
    expect(wrapper.find(StatementSVG)).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });
  it('should pass custom props to correct blocks', () => {
    const wrapper = shallow(
      <ModifiersList
        items={items}
        sizesTableUrl='test.ru/123'
        wrapperProps={{
          className: 'wrapper-test-class',
          id: 4,
        }}
        itemsContainerProps={{
          className: 'items-container-test-class',
        }}
      />,
    );

    const wrapperElement = wrapper.find('.wrapper-test-class');
    expect(wrapperElement.prop('id')).toEqual(4);
    expect(wrapperElement.find(Link)).toHaveLength(1);
    expect(wrapperElement.find(ModifierItem)).toHaveLength(2);

    const itemsContainerElement = wrapper.find('.items-container-test-class');
    expect(itemsContainerElement.find(Link)).toHaveLength(0);
    expect(itemsContainerElement.find(ModifierItem)).toHaveLength(2);
    expect(wrapper).toMatchSnapshot();
  });
  it('should call onClickItem correctly on click ModifierItem', () => {
    const onItemClickFn = jest.fn();
    const wrapper = mount(<ModifiersList items={items} onItemClick={onItemClickFn} />);

    expect(onItemClickFn).not.toHaveBeenCalled();
    wrapper.find(ModifierItem).at(1).simulate('click');
    expect(onItemClickFn).toHaveBeenCalledTimes(1);
    expect(onItemClickFn).toHaveBeenCalledWith(items[1]);

    wrapper.find(ModifierItem).at(0).simulate('click');
    expect(onItemClickFn).toHaveBeenCalledTimes(2);
    expect(onItemClickFn).toHaveBeenNthCalledWith(2, items[0]);
  });
});
