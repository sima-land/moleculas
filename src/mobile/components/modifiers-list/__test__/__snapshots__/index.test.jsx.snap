// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModifiersList /> should pass custom props to correct blocks 1`] = `
<div
  className="wrapper-test-class"
  id={4}
>
  <div
    className="sizes-table-wrapper"
  >
    <ForwardRef(Link)
      className="sizes-table-link"
      href="test.ru/123"
    >
      Таблица размеров
      <ForwardRef(StatementSVG)
        className="sizes-table-icon"
        role="presentation"
      />
    </ForwardRef(Link)>
  </div>
  <div
    className="items-container-test-class"
  >
    <ModifierItem
      key="0"
      name="testModifier - 1"
      onClick={[Function]}
      price={123}
      type="text"
    />
    <ModifierItem
      key="1"
      name="testModifier - 2"
      onClick={[Function]}
      price={233}
      type="text"
    />
  </div>
</div>
`;

exports[`<ModifiersList /> should render correct with sizesTableUrl 1`] = `
<ModifiersList
  sizesTableUrl="test.ru/123"
>
  <div>
    <div
      className="sizes-table-wrapper"
    >
      <ForwardRef(Link)
        className="sizes-table-link"
        href="test.ru/123"
      >
        <a
          className="link sizes-table-link color__brand-blue hover-color__brand-deep-blue"
          href="test.ru/123"
        >
          Таблица размеров
          <ForwardRef(StatementSVG)
            className="sizes-table-icon"
            role="presentation"
          >
            <svg
              className="sizes-table-icon"
              height={24}
              role="presentation"
              viewBox="0 0 24 24"
              width={24}
            >
              <path
                d="M8 16a1 1 0 001 1h4a1 1 0 100-2H9a1 1 0 00-1 1zM9 13a1 1 0 110-2h6a1 1 0 110 2H9zM8 8a1 1 0 001 1h6a1 1 0 100-2H9a1 1 0 00-1 1z"
              />
              <path
                clipRule="evenodd"
                d="M6 2h12a3 3 0 013 3v14a3 3 0 01-3 3H6a3 3 0 01-3-3V5a3 3 0 013-3zm0 2a1 1 0 00-1 1v14a1 1 0 001 1h12a1 1 0 001-1V5a1 1 0 00-1-1H6z"
                fillRule="evenodd"
              />
            </svg>
          </ForwardRef(StatementSVG)>
        </a>
      </ForwardRef(Link)>
    </div>
  </div>
</ModifiersList>
`;

exports[`<ModifiersList /> should renders correctly with correct props 1`] = `
<div>
  <div>
    <ModifierItem
      currencyGrapheme="$"
      key="0"
      name="testModifier - 1"
      onClick={[Function]}
      price={123}
      type="text"
    />
    <ModifierItem
      currencyGrapheme="$"
      key="1"
      name="testModifier - 2"
      onClick={[Function]}
      price={233}
      type="text"
    />
  </div>
</div>
`;

exports[`<ModifiersList /> should renders correctly without props 1`] = `<div />`;
